firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -182,16 +187,46 @@ private void startMutationQueue() {
   /** Accepts locally generated Mutations and commits them to storage. */
   public LocalWriteResult writeLocally(List<Mutation> mutations) {
     Timestamp localWriteTime = Timestamp.now();
+
     // TODO: Call queryEngine.handleDocumentChange() appropriately.
-    MutationBatch batch =
-        persistence.runTransaction(
-            "Locally write mutations",
-            () -> mutationQueue.addMutationBatch(localWriteTime, mutations));
 
-    Set<DocumentKey> keys = batch.getKeys();
-    ImmutableSortedMap<DocumentKey, MaybeDocument> changedDocuments =
+    Set<DocumentKey> keys = new HashSet<>();
+    for (Mutation mutation : mutations) {
+      keys.add(mutation.getKey());
+    }
+
+    // Load and apply all existing mutations. This lets us compute the current base state for all
+    // non-idempotent transforms before applying any additional user-provided writes.
+    ImmutableSortedMap<DocumentKey, MaybeDocument> existingDocuments =
         localDocuments.getDocuments(keys);
-    return new LocalWriteResult(batch.getBatchId(), changedDocuments);
+
+    return persistence.runTransaction(
+        "Locally write mutations",
+        () -> {
+
+          // For non-idempotent mutations (such as `FieldValue.numericAdd()`), we record the base
+          // state in a separate patch mutation. This is later used to guarantee consistent values
+          // and prevents flicker even if the backend sends us an update that already includes our
+          // transform.
+          List<Mutation> baseMutations = new ArrayList<>();
+          for (Mutation mutation : mutations) {
+            MaybeDocument maybeDocument = existingDocuments.get(mutation.getKey());
+            if (maybeDocument instanceof Document && !mutation.isIdempotent()) {
```

Comment:
What is the latency compensation behavior of:
```
db.collection('votes/michael').set({count: FieldValue.numericAdd(1)}, {merge: true})
```
given votes/michael is not already in cache?

I think I'd expect to get a latency compensated event with {count: 1} in it (do we?).  Assuming we do, I think this code as written won't write a baseValue for 'count' (since we don't have one).  And so it'll be subject to flicker if you receive the watch update before the write ack.

That said, in order for you not to have votes/michael in cache, you can't have already listened to it. So perhaps the only way this could come up is if you start listening while offline (so we'll get latency compensation but it won't be in cache yet), do the merge-set increment, and then come online, and get the watch update before the write ack.  Maybe this is enough of an edge case we don't care that much.

Thoughts?  I can't think of an easy way to fix this unfortunately.  If we care, we may need to re-think the baseMutations thing.

Reply:
LOL 😛 

But now what if instead of a merge-set you do an update() when you have no cached document?

  db.collection('votes/michael').update({count: FieldValue.numericAdd(1)})

I think this will now show up in latency compensation, when it should not (an update() without a base document doesn't get applied to the local view since we don't want to present you with a partial document unless you've specifically opted into this by doing a merge-set instead).

So I'm not sure your fix is right...  I think to get this right we'd only create baseMutations if a) we have a base document, or b) there's a Precondition=None mutation prior to the non-idempotent mutation...  or something like that.

Reply:
I improved my fix further by re-using the same Precondition in the base mutation. I think that should work... Hopefully. Otherwise I have to go back to the drawing board.